---
title: "Bioconductor Part 1"
author: "Kirsten Gotting"
date: "April 12, 2018"
output: html_document
---

"Bioconductor provides core data structures and methods that enable genome-scale analysis of high-throughput data in the context of the rich statistical programming environment offered by the R project2."(1)


"Bioconductor documentation comes at three levels: workflows ... package vignettes ...  and function manual pages. In many cases, users ultimately become developers, making their own algorithms and approaches available to others." (1)

Key Packages:

IRanges

GenomicRanges

GenomicAlignments

GenomicFeatures

VariantAnnotation

rtracklayer

AnnotationHub

```{r, include=FALSE}

#source("https://bioconductor.org/biocLite.R")
packages <- c("IRanges", 
              "GenomicRanges", 
              "GenomicAlignments", 
              "GenomicFeatures", 
              "VariantAnnotation", 
              "rtracklayer", 
              "AnnotationHub")

#biocLite(pkgs=packages)

lapply(packages, library, character.only = TRUE)

```

# GenomicRanges

```{r}
help(package='IRanges')
help(package='GenomicAlignments')


## GRanges object:
gr <- GRanges(
        seqnames=Rle(c("chr1", "chr2", "chr1", "chr3"), c(1, 3, 2, 4)),
        ranges=IRanges(1:10, width=10:1, names=head(letters,10)),
        strand=Rle(strand(c("-", "+", "*", "+", "-")), c(1, 2, 2, 3, 2)),
        score=1:10,
        GC=seq(1, 0, length=10)
      )
gr

## GRangesList object:
gr1 <- GRanges(seqnames="chr2", ranges=IRanges(4:3, 6),
               strand="+", score=5:4, GC=0.45)
gr1

## finding overlapping intervals in different ways (similar to the python package HTseq)

##  countOverlaps counts the number of subject ranges that overlap with the query range
countOverlaps(query = gr, subject = gr1)


## subsetByOverlaps returns the subset of x that has an overlap hit with a range in ranges using the specified findOverlaps parameters
subsetByOverlaps(gr, gr1)


gff_file <- system.file("extdata", "GFF3_files", "a.gff3", package="GenomicFeatures")
## see ?rtracklayer::import.gff3 to import your own annotation date
## and ?rtracklayer::import for importing different data types (BED, bedGraph, WIG, BigWig)

txdb     <- makeTxDbFromGFF(gff_file, format="gff3")
txdb 

exons <- exonsBy(txdb, by="gene")
exons

```

# SummarizedExperiment

```{r}

nrows <- 200
ncols <- 6

## create a matrix of fake counts
counts <- matrix(runif(nrows * ncols, 1, 1e4), nrows)
head(counts) ## columns are samples, rows are genes


## create a fake annotation object (ie import.gff3)
rowRanges <- GRanges(rep(c("chr1", "chr2"), c(50, 150)),
                     IRanges(floor(runif(200, 1e5, 1e6)), width=100),
                     strand=sample(c("+", "-"), 200, TRUE),
                     feature_id=sprintf("ID%03d", 1:200))

rowRanges

## create a fake 'colData' object with the information about your replicates/treatments

colData <- DataFrame(Treatment=rep(c("ChIP", "Input"), 3),
                     row.names=LETTERS[1:6])

rse <- SummarizedExperiment(assays=SimpleList(counts=counts),
                            rowRanges=rowRanges, 
                            colData=colData)

head(assay(rse))

colData(rse)
rowData(rse)

```

# AnnotationHub

```{r}


ah <- AnnotationHub()
ah

unique(ah$dataprovider)

unique(ah$rdataclass)

orgs <- subset(ah, ah$rdataclass == "OrgDb")
orgs


dog <- query(orgs, "Canis familiaris")[[1]]

columns(dog)

all_keys <- keys(dog, keytype="ACCNUM")


select(dog, 
       keys=head(all_keys), 
       columns=c('SYMBOL', 'ALIAS', 'UNIPROT', 'GENENAME', 'PMID'), 
       keytype="ACCNUM")



ranges <- subset(ah, ah$rdataclass == "GRanges")
ranges


dog <- query(ranges, "Canis familiaris")[[1]]



```

# References
1. Huber, W., Carey, V. J., Gentleman, R., Anders, S., Carlson, M., Carvalho, B. S., ... & Gottardo, R. (2015). Orchestrating high-throughput genomic analysis with Bioconductor. Nature methods, 12(2), 115.
Link: https://www.nature.com/articles/nmeth.3252


# Relevant Vignettes
[Counting reads with summarizeOverlaps](https://bioconductor.org/packages/release/bioc/vignettes/GenomicAlignments/inst/doc/summarizeOverlaps.pdf)

[Overlap Encodings](https://www.bioconductor.org/packages/release/bioc/vignettes/GenomicAlignments/inst/doc/OverlapEncodings.pdf)
